### PR TITLE
Example simple wallet with async multisig support.

### PR DESCRIPTION
**Pull request by dmkozh**
### What

Simple smart wallet example (https://github.com/stellar/soroban-examples/issues/73).

The wallet supports weighted multisig with arbitrarily separated signature batches (i.e. the payment may happen both immediately or after a sequence of confirmations from signers). 

### Why

With this example I've tried to build multisig contract using the existing SDK tools. This should be useful as we keep iterating on auth, signing utils etc.

### Known limitations

Some book-keeping (like payment id generation) is not implemented in the contract itself, though that doesn't prevent it from being correct/functional.

### CODE REVIEW

File: wallet/src/lib.rs
Comment (leighmcculloch) [inline wallet/src/lib.rs:113]: We should encourage including a domain separator in anything that gets signed, so we should probably add it to this example, if we can find a concise way to do so.

File: wallet/Cargo.toml
Comment (leighmcculloch) [inline wallet/Cargo.toml:19]: Before we can merge this contract we need to update how it interacts with tokens by importing the `soroban-token-spec.wasm` for the client, and creating and using the built-in token contract.

**Review by leighmcculloch — COMMENTED** [review]
This is very nice example. For something that has a certain depth of complexity, the example presents that complexity in a way that is rather brilliant in its simplicity.

There's one thing I mention inline I think we should tweak so as to encourage a good security practice. There's one thing I mention inline we need to update just because it is out-of-date.

I'm happy to help update the contract, but defer to you @dmkozh if you want to do it.

cc @tomerweller if you haven't seen this contract, it is worth a read.

File: wallet/Cargo.toml
Comment (dmkozh) [reply wallet/Cargo.toml]: Updated the contract to match the new SDK and token contract.

File: wallet/src/lib.rs
Comment (dmkozh) [reply wallet/src/lib.rs]: I'm not sure what do you mean by the domain separator here. Network id is a part of the signature payload expected by `verify`. Do you mean something else here?

Comment (leighmcculloch) [reply wallet/src/lib.rs]: Ok, I missed that this section was using soroban-auth. As long as it's using the signature payload from that crate then it's good.

File: wallet/src/lib.rs
Comment (leighmcculloch) [inline wallet/src/lib.rs:178]: Given the amount of input validation this contract would benefit from using `panic_error!` I think, so that if there are panics in the final compiled, there are contract error codes rather than the trap error code. See https://soroban-dev.stellar.org/docs/examples/errors#panicking-with-an-error.

**Review by leighmcculloch — APPROVED** [review]

File: wallet/src/lib.rs
Comment (dmkozh) [reply wallet/src/lib.rs:178]: I agree that errors would be nice. Moreover, I would probably rewrite this to return `Result` instead of panicking and use `try_` functions in tests as `should_panic` is pretty annoying. The only thing I'm not sure about is granularity: replacing every panic with a unique error is a bit annoying (and also inconvenient for a potential consumer of these errors), while reducing granularity would lose some information/test coverage (now I can precisely set expectations on why should the contract fail). So ideally I'd want to both return an error and have a string explanation (also as a debug event?) for debugging/testing.

Comment (leighmcculloch) [reply wallet/src/lib.rs:178]: I'd lean into granularity.

You can still use the `try` functions if you use the `panic_error!` macro. This might be a good balance if the intent is mostly for testing and diagnostics, and less for developers to handle them explicitly. The only difference between returning Result or not is that the Error gets encoded into the contract spec. So if you want people to handle errors, getting it into the contract spec helps to document it.

> So ideally I'd want to both return an error and have a string explanation (also as a debug event?) for debugging/testing.

We might want to explore modifying `panic_error!` so that it can optionally take a string as well, and it can log a debug event using `log!` alongside storing the error.

Comment (dmkozh) [reply wallet/src/lib.rs:178]: > I'd lean into granularity.

True, I guess unlike with host more granular errors are more acceptable for contracts (given that they're immutable). 

> We might want to explore modifying panic_error! so that it can optionally take a string as well, and it can log a debug event using log! alongside storing the error.

That's what I've been thinking about, however currently the debug events from `log` don't seem to appear in the test logs, so we need to think how to make this useful. For now I'll just use granular errors.

Comment (leighmcculloch) [reply wallet/src/lib.rs:178]: You can make the logs appear in test logs by adding this line to the end of your test:
```rust
env.logger().print()
```

I'd love to find a way to hook this in automatically if you have any idea.

Comment (dmkozh) [reply wallet/src/lib.rs:178]: > The only difference between returning Result or not is that the Error gets encoded into the contract spec.

It seems like another different thing is that `panic_error` still panics, which makes tests harder (and IIRC we can't `catch_unwind` that). So for the sake of testing `Result` seems preferential, as `should_panic` is inconvenient when there are a lot of failure scenarios to cover.

Comment (leighmcculloch) [reply wallet/src/lib.rs:178]: > `panic_error` still panics, which makes tests harder

Agreed, and there's an issue open to address this: 
- https://github.com/stellar/rs-soroban-sdk/issues/642
- https://github.com/stellar/rs-soroban-env/issues/430

Comment (leighmcculloch) [reply wallet/src/lib.rs:178]: Fwiw I think returning Results from these functions would also be very reasonable. I'm really curious if you add that how much the contract size changes. I think we've been worried return paths could be expensive, but it's all assumptions at this point, and you might prove that wrong, which imo would be awesome. So awesome we could maybe even axe panic_error!.

Comment (dmkozh) [reply wallet/src/lib.rs:178]: In https://github.com/stellar/soroban-examples/pull/150 I'm using results everywhere besides `ConversionError` unwraps. I think using result doesn't make things much more complex/verbose, It definitely is beneficial for testing and the code is IMHO more idiomatic Rust.

**Pull request merged**
